lib/driver-cli.js

```diff
@@ -0,0 +1,249 @@
+          typeTxt = '(NPM)';
+      }
+      const installTxt = installed ?
+        `@${version.yellow} ${('[installed ' + typeTxt + ']').green}` :
+        ' [not installed]'.grey;
+      const updateTxt = showUpdates && updateVersion ?
+        ` [${updateVersion} available]`.magenta :
+        '';
+
+      console.log(`- ${name.yellow}${installTxt}${updateTxt}`); // eslint-disable-line no-console
+    }
+
+    return drivers;
+  }
+
+  async install ({driver, installType}) {
+    log(this.json, `Attempting to find and install driver '${driver}'`);
+
+    let driverData, installSpec = driver;
+
+    if (installType === INSTALL_TYPE_LOCAL) {
+      const msg = 'Linking driver from local path';
+      const pkgJsonData = await spinWith(this.json, msg, async () => (
+        await this.npm.linkPackage(installSpec))
+      );
+      driverData = this.getDriverFields(pkgJsonData);
+      driverData.installPath = driverData.pkgName;
+    } else if (installType === INSTALL_TYPE_GITHUB) {
+      if (installSpec.split('/').length !== 2) {
+        throw new Error(`Github driver spec ${installSpec} appeared to be invalid; ` +
+                        'it should be of the form <org>/<repo>');
+      }
+      driverData = await this.installViaNpm({driver: installSpec});
+    } else if (installType === INSTALL_TYPE_GIT) {
+      driverData = await this.installViaNpm({driver: installSpec});
+    } else {
+      // at this point we have either an npm package or an appium verified driver
+      // name. both of which will be installed via npm.
+      // drivers installed via npm can include versions or tags after the '@'
+      // sign, so check for that
+      const [name, pkgVer] = installSpec.split('@');
+      let pkgName;
+
+      if (installType === INSTALL_TYPE_NPM) {
+        // if we're installing a named package from npm, we don't need to check
+        // against the appium driver list; just use the installSpec as is
+        pkgName = name;
+      } else {
+        // if we're installing a named appium driver (like 'xcuitest') we need to
+        // dereference the actual npm package ('appiupm-xcuitest-driver'), so
+        // check it exists and get the correct package
+        const knownNames = Object.keys(KNOWN_DRIVERS);
+        if (!_.includes(knownNames, name)) {
+          const msg = `Could not resolve driver; are you sure it's in the list ` +
+                      `of supported drivers? ${JSON.stringify(knownNames)}`;
+          throw new Error(msg);
+        }
+        pkgName = KNOWN_DRIVERS[name];
+        // given that we'll use the install type in the driver json, store it as
+        // 'npm' now
+        installType = INSTALL_TYPE_NPM;
+      }
+
+      driverData = await this.installViaNpm({driver, pkgName, pkgVer});
+    }
+
+    const {driverName} = driverData;
+    delete driverData.driverName;
+
+    if (this.config.isInstalled(driverName)) {
+      throw new Error(`A driver named '${driverName}' is already installed. ` +
```

Comment:
should this really be an error?

Reply:
i tend to think so, especially since we might be in a json context where there's no other log output. if something doesn't happen the way the user expects i would take it to be an error.